packages/rpc/src/chainStorageWatcher.ts

```diff
@@ -51,136 +58,101 @@ export const makeAgoricChainStorageWatcher = (
   chainId: string,
   onError?: (e: Error) => void,
   marshaller = makeClientMarshaller(),
-  newPathQueryDelayMs = defaults.newPathQueryDelayMs,
   refreshLowerBoundMs = defaults.refreshLowerBoundMs,
   refreshUpperBoundMs = defaults.refreshUpperBoundMs,
 ) => {
-  // Map of paths to [identifier, value] pairs of most recent response values.
-  //
-  // The 'identifier' is used to avoid notifying subscribers for already-seen
-  // values. For 'data' queries, 'identifier' is the blockheight of the
-  // response. For 'children' queries, 'identifier' is the stringified array
-  // of children.
+  /**
+   * Map from vstorage paths to `[identifier, value]` pairs containing their
+   * most recent response values and their {@link LatestValueIdentifier}.
+   */
   const latestValueCache = new Map<
     string,
-    [identifier: string, value: unknown]
+    [identifier: LatestValueIdentifier, value: unknown]
   >();
 
   const watchedPathsToSubscribers = new Map<string, Set<Subscriber<unknown>>>();
-  let isNewPathWatched = false;
-  let isQueryInProgress = false;
-  let nextQueryTimeout: number | null = null;
-
-  const queueNextQuery = () => {
-    if (isQueryInProgress || !watchedPathsToSubscribers.size) {
-      return;
-    }
+  const watchedPathsToRefreshTimeouts = new Map<string, number>();
 
-    if (isNewPathWatched) {
-      // If there is any new path to watch, schedule another query very soon.
-      if (nextQueryTimeout) {
-        window.clearTimeout(nextQueryTimeout);
-      }
-      nextQueryTimeout = window.setTimeout(queryUpdates, newPathQueryDelayMs);
-    } else {
-      // Otherwise, refresh after a normal interval.
-      nextQueryTimeout = window.setTimeout(
-        queryUpdates,
+  const refreshDataForPath = async (
+    path: [AgoricChainStoragePathKind, string],
+  ) => {
+    const queueNextRefresh = () => {
+      window.clearTimeout(watchedPathsToRefreshTimeouts.get(pathKey));
+      const timeout = window.setTimeout(
+        () => refreshDataForPath(path),
         randomRefreshPeriod(refreshLowerBoundMs, refreshUpperBoundMs),
       );
-    }
-  };
+      watchedPathsToRefreshTimeouts.set(pathKey, timeout);
+    };
 
-  const queryUpdates = async () => {
-    isQueryInProgress = true;
-    nextQueryTimeout = null;
-    isNewPathWatched = false;
-
-    const paths = [...watchedPathsToSubscribers.keys()].map(keyToPath);
-
-    if (!paths.length) {
-      isQueryInProgress = false;
+    const pathKey = pathToKey(path);
+    let response;
+    try {
+      response = await vstorageQuery(apiAddr, marshaller.fromCapData, path);
+    } catch (e: unknown) {
+      console.error(`Error querying vstorage for path ${path}:`, e);
+      if (e instanceof Error) {
+        onError && onError(e);
+      }
```

Comment:
consider one conditional expression,
```suggestion
      if (onError && e instanceof Error) {
        onError(e);
      }
```

Reply:
Done